feat: Upgrade OTel Collector to version 0.120.0 (contrib 0.120.1) (#1873)
kyma-project/telemetry-manager: docs/user/resources/01-telemetry.md

For details, see the [Telemetry specification file](https://github.com/kyma-project/telemetry-manager/blob/main/apis/operator/v1alpha1/telemetry_types.go).

### Annotations

> The OpenTelemetry internal metric backward compatibility:

Reviewer: ```suggestion
backward compatibility for internal metrics in OpenTelemetry:
```

> The recent version of OpenTelemetry Collector 0.120.0 introduced a new suffix for internal metrics exposed through the Prometheus endpoint and breaks the stability of internal metrics.

Reviewer 2: It is also about the scope and the unit, maybe provide a concrete example?

For details, see the [Telemetry specification file](https://github.com/kyma-project/telemetry-manager/blob/main/apis/operator/v1alpha1/telemetry_types.go).

### Annotations

Backward compatibility for internal metrics in OpenTelemetry:

> The recent version of OpenTelemetry Collector 0.119.0 introduced breaking changes for internal metrics exposed through the Prometheus endpoint and breaks the stability of internal metrics.

Reviewer 3: We won't remember to remove the "recent" bit here even when it's not recent anymore. We can simply drop it now.

```suggestion
OpenTelemetry Collector 0.119.0 introduces breaking changes for internal metrics exposed through the Prometheus endpoint and breaks the stability of internal metrics. This affects Kyma Telemetry in the following ways:
```

> - The metric name, for example, the counter metrics will be appending a `_total` suffix

Reviewer 3: No sure what this means, here's my guess:

```suggestion
- The metric name changes. For example, the counter metrics append a `_total` suffix.
```

> - The metric unit will be appended to the metric name, for example, a counter metric `request_duration` with unit `milliseconds` will be exposed as `request_duration_milliseconds_total`

Reviewer 3: ```suggestion
- The metric unit is appended to the metric name. For example, a counter metric `request_duration` with unit `milliseconds` is exposed as `request_duration_milliseconds_total`.
```

> - The internal metric exporter will create a `otel_scope_info` metric containing the metrics Instrumentation Scope, and also add labels about Instrumentation Scope to all metric points.

Reviewer 3: ```suggestion
- The internal metric exporter creates an `otel_scope_info` metric containing the metrics instrumentation scope, and also add labels about instrumentation scope to all metric points.
```

I don't think [instrumentation scope](https://opentelemetry.io/docs/concepts/instrumentation-scope/) needs capitalization.

> To maintain backward compatibility, the Telemetry Manager introduces new annotations `telemetry.kyma-project.io/internal-metrics-compatibility-mode` to control the internal metrics suffix.

Reviewer 3: - "introduces" implies they are new.
- Is it one or more annotation?

```suggestion
To maintain backward compatibility, the Telemetry Manager introduces the annotation `telemetry.kyma-project.io/internal-metrics-compatibility-mode` to control the internal metrics suffix.
```

To enable the backward compatibility mode, set the annotation `telemetry.kyma-project.io/internal-metrics-compatibility-mode: true` in the Telemetry CR.
```yaml
apiVersion: operator.kyma-project.io/v1alpha1
kind: Telemetry
metadata:
  name: default
  namespace: kyma-system
  annotations:
    telemetry.kyma-project.io/internal-metrics-compatibility-mode: "true"
```

> [! WARNING]

> > The backward compatibility mode is disabled by default. This annotation will be removed in the future.

Reviewer 2: ```suggestion
> The backward compatibility mode is disabled by default. This annotation will be removed in version 1.41.0.
```

Reviewer 3: [veto](https://github.com/kyma-project/telemetry-manager/pull/1873#discussion_r1977173492)

> > The backward compatibility mode is disabled by default. This annotation will be removed in version 1.41.0.

Reviewer 3: ```suggestion
> By default, the backward compatibility mode is disabled.```

We don't announce future plans in our docs. That's what release notes and deprecation notes are for.
